File: Econovation_Algorithm_Mentoring/7. 그래프의 탐색 (DFS, BFS)/Programmers_Network.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;

public class Programmers_Network {
    boolean[] visit;
    ArrayList<ArrayList<Integer>> list;
    static int count;

    public int solution(int n, int[][] computers) {
        int answer = 0;
        visit = new boolean[n];
        list = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            list.add(new ArrayList<>());
        }
        for (int i = 0; i < computers.length; i++) {
            for (int j = 0; j < computers[0].length; j++) {
                if (computers[i][j] == 1) {
                    list.get(i).add(j);
                }
            }
        }
        count = 0;
        for(int i=0;i<n;i++){
            if(!visit[i]){
                bfs(i);
            }
        }
        answer = count;

        return answer;
    }

    public void bfs(int n) {
        Queue<Integer> q = new LinkedList<>();
        q.add(n);
        visit[n] = true;

        while (!q.isEmpty()) {
            int x = q.poll();
            for (int i = 0; i < list.get(x).size(); i++) {
                if (!visit[list.get(x).get(i)]) {
                    q.add(list.get(x).get(i));
                    visit[list.get(x).get(i)] = true;
                }
            }

        }
        count++;
    }
}

File: Algorithm_Swift/Algorithm_Swift/Programmers/Programmers_Network.swift
//
//  Programmers_Network.swift
//  Algorithm_Swift
//
//  Created by asong on 2022/03/24.
//

import Foundation

func Programmers_Network(){
    var visit: [Bool] = []
    var list: [[Int]] = []
    var answer = 0
    func solution(_ n:Int, _ computers:[[Int]]) -> Int {
        visit = Array(repeating:false, count: n)
        list = Array(repeating: [], count: n)
        for i in 0..<computers.count{
            for j in 0..<computers[0].count{
                if computers[i][j] == 1 && i != j{
                    list[i].append(j)
                }
            }
        }
        for i in 0..<computers.count{
            if !visit[i]{
                bfs(i)
            }
        }
        return answer
    }
    func bfs(_ n: Int){
        visit[n] = true
        var q = [n]
        while !q.isEmpty{
            let now = q.removeFirst()
            for i in 0..<list[now].count{
                if !visit[list[now][i]]{
                    q.append(list[now][i])
                    visit[list[now][i]] = true
                }
            }
        }
        answer += 1
    }
}
